### ISSUE TITLE
<<loop>> when using shown twice

### ISSUE THREAD

**Issue by ocharles** [chrisdone/formatting#36]
The following works fine:

```haskell
formatCoordinate =
  Formatting.shown
```

However, the following causes a loop:

```haskell
formatCoordinate =
  Formatting.shown <> Formatting.shown
```

I'd expect that to print the same thing twice, with a comma between. No idea what's causing a loop, there doesn't seem to be any obvious recursion in the source code for `<>`!

**Comment by ocharles** [comment]
Minimal reproducible example:

```
Prelude Formatting> import Data.Monoid
Prelude Formatting Data.Monoid> bprint (shown <> shown) True
"
```

**Comment by ocharles** [comment]
```
Prelude> import Data.Semigroup 
Prelude Data.Semigroup> import Formatting
Prelude Data.Semigroup Formatting> bprint (shown <> shown) True
"TrueTrue"
```

Looks like for me mappend is defined as Data.Monoid.<>, which causes the loop.

**Comment by chrisdone** [comment]
Right, the source is:

```haskell
instance Sem.Semigroup (Format r (a -> r)) where
  (<>) m n =
    Format (\k a ->
              runFormat m (\b1 -> runFormat n (\b2 -> k (b1 <> b2)) a) a)

instance Monoid (Format r (a -> r)) where
  mappend = (<>)
  mempty = Format (\k _ -> k mempty)
```

The `(<>)` comes from the `Semigroup` instance. I'm not sure what's going on, I'll load formatting in on GHC 8. My stack.yaml is currently on GHC 7.10.

**Comment by chrisdone** [comment]
OK, I reproduce on GHC 7.10 too:

```haskell
> format (later id <> later id) "x"
"  c cInterrupted.
```

**Comment by chrisdone** [comment]
OK, this regression was introduced in 9112eab27628141774f55c1463fbb2e51c25f59f

I'll revert it and explore a different support for 8.4.

**Comment by chrisdone** [comment]
Fixed and bumped in 6.3.2. Cheers!
